vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v3.cc

```diff
@@ -0,0 +1,118 @@
+#include "bat/ads/internal/base/logging_util.h"
+#include "bat/ads/internal/creatives/new_tab_page_ads/creative_new_tab_page_ads_database_table.h"
+#include "bat/ads/internal/geographic/subdivision/subdivision_targeting.h"
+#include "bat/ads/internal/resources/behavioral/anti_targeting/anti_targeting_resource.h"
+
+namespace ads::new_tab_page_ads {
+
+EligibleAdsV3::EligibleAdsV3(
+    geographic::SubdivisionTargeting* subdivision_targeting,
+    resource::AntiTargeting* anti_targeting_resource)
+    : EligibleAdsBase(subdivision_targeting, anti_targeting_resource) {}
+
+void EligibleAdsV3::GetForUserModel(
+    const targeting::UserModelInfo& user_model,
+    GetEligibleAdsCallback<CreativeNewTabPageAdList> callback) {
+  BLOG(1, "Get eligible new tab page ads");
+
+  database::table::AdEvents database_table;
+  database_table.GetForType(
+      mojom::AdType::kNewTabPageAd,
+      [=](const bool success, const AdEventList& ad_events) {
+        if (!success) {
+          BLOG(1, "Failed to get ad events");
+          callback(/*had_opportunity*/ false, {});
+          return;
+        }
+
+        GetBrowsingHistory(user_model, ad_events, callback);
+      });
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+void EligibleAdsV3::GetBrowsingHistory(
+    const targeting::UserModelInfo& user_model,
+    const AdEventList& ad_events,
+    GetEligibleAdsCallback<CreativeNewTabPageAdList> callback) {
+  const int max_count = features::GetBrowsingHistoryMaxCount();
+  const int days_ago = features::GetBrowsingHistoryDaysAgo();
+  AdsClientHelper::GetInstance()->GetBrowsingHistory(
+      max_count, days_ago,
+      base::BindOnce(&EligibleAdsV3::GetEligibleAds, base::Unretained(this),
+                     user_model, ad_events, callback));
+}
+
+void EligibleAdsV3::GetEligibleAds(
+    const targeting::UserModelInfo& user_model,
+    const AdEventList& ad_events,
+    GetEligibleAdsCallback<CreativeNewTabPageAdList> callback,
+    const BrowsingHistoryList& browsing_history) {
+  database::table::CreativeNewTabPageAds database_table;
+  database_table.GetAll([=](const bool success, const SegmentList& /*segments*/,
+                            const CreativeNewTabPageAdList& creative_ads) {
+    if (!success) {
+      BLOG(1, "Failed to get ads");
+      callback(/*had_opportunity*/ false, {});
+      return;
+    }
+
+    if (creative_ads.empty()) {
+      BLOG(1, "No eligible ads");
+      callback(/*had_opportunity*/ false, {});
+      return;
+    }
+
+    const CreativeNewTabPageAdList eligible_creative_ads =
+        FilterCreativeAds(creative_ads, ad_events, browsing_history);
+    if (eligible_creative_ads.empty()) {
+      BLOG(1, "No eligible ads out of " << creative_ads.size() << " ads");
+      callback(/*had_opportunity*/ true, {});
+      return;
+    }
+
+    const absl::optional<CreativeNewTabPageAdInfo> creative_ad =
+        PredictAd(user_model, ad_events, eligible_creative_ads);
```

Comment:
Not a comment on this PR, but we should create an issue for changing the non-notification ad types to use embedding-based v3 predicting